File: packages/typescript/src/formatter.ts
import { AstNode, BlockNode, DocumentNode } from "./ast.js";

function q(s: string): string {
  return `"${s.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n").replace(/\t/g, "\\t")}"`;
}

function formatValue(node: AstNode): string {
  if (node.kind === "string") return q(node.value);
  if (node.kind === "number") return `${node.value}`;
  if (node.kind === "boolean") return node.value ? "true" : "false";
  if (node.kind === "array") return `[${node.elements.map(formatValue).join(", ")}]`;
  throw new Error(`unsupported value kind: ${node.kind}`);
}

function formatBlock(block: BlockNode, indent = 0): string {
  const pad = "  ".repeat(indent);
  const head = block.argument ? `${pad}${block.name} ${q(block.argument)} do` : `${pad}${block.name} do`;
  const out: string[] = [head];

  for (const [k, v] of Object.entries(block.properties)) out.push(`${pad}  ${k} = ${formatValue(v)}`);

  const seen = new Set<BlockNode>();
  for (const child of Object.values(block.blocks)) {
    if (seen.has(child)) continue;
    seen.add(child);
    out.push(formatBlock(child, indent + 1));
  }
  for (const child of block.named_blocks ?? []) {
    if (seen.has(child)) continue;
    seen.add(child);
    out.push(formatBlock(child, indent + 1));
  }

  out.push(`${pad}end`);
  return out.join("\n");
}

export function format(document: DocumentNode): string {
  return document.blocks.map((b) => formatBlock(b)).join("\n\n");
}

File: packages/rust/src/formatter.rs
use crate::ast::{AstNode, BlockNode, DocumentNode};
use std::collections::BTreeSet;

pub fn format(doc: &DocumentNode) -> String {
    doc.blocks.iter().map(|b| format_block(b, 0)).collect::<Vec<_>>().join("\n\n")
}

fn q(s: &str) -> String {
    format!(
        "\"{}\"",
        s.replace('\\', "\\\\").replace('"', "\\\"").replace('\n', "\\n").replace('\t', "\\t")
    )
}

fn format_value(v: &AstNode) -> String {
    match v {
        AstNode::String { value, .. } => q(value),
        AstNode::Number { value, .. } => value.to_string(),
        AstNode::Boolean { value, .. } => if *value { "true".into() } else { "false".into() },
        AstNode::Array { elements, .. } => format!("[{}]", elements.iter().map(format_value).collect::<Vec<_>>().join(", ")),
    }
}

fn format_block(b: &BlockNode, indent: usize) -> String {
    let pad = "  ".repeat(indent);
    let mut lines = vec![if let Some(arg) = &b.argument { format!("{}{} {} do", pad, b.name, q(arg)) } else { format!("{}{} do", pad, b.name) }];

    for (k, v) in &b.properties { lines.push(format!("{}  {} = {}", pad, k, format_value(v))); }

    let mut seen = BTreeSet::new();
    for child in b.blocks.values() {
        let key = if let Some(arg) = &child.argument { format!("{}:{}", child.name, arg) } else { child.name.clone() };
        if seen.insert(key) { lines.push(format_block(child, indent + 1)); }
    }
    for child in &b.named_blocks {
        let key = if let Some(arg) = &child.argument { format!("{}:{}", child.name, arg) } else { child.name.clone() };
        if seen.insert(key) { lines.push(format_block(child, indent + 1)); }
    }

    lines.push(format!("{}end", pad));
    lines.join("\n")
}
